cabal-install/src/Distribution/Client/HttpUtils.hs

```diff
@@ -535,12 +536,13 @@ curlTransport prog =
             (Just (URIAuth u _ _)) | not (null u) -> Just $ filter (/= '@') u
             _ -> Nothing
       -- prefer passed in auth to auth derived from uri. If neither exist, then no auth
-      let mbAuthString = case (explicitAuth, uriDerivedAuth) of
-            (Just (uname, passwd), _) -> Just (uname ++ ":" ++ passwd)
-            (Nothing, Just a) -> Just a
-            (Nothing, Nothing) -> Nothing
-      case mbAuthString of
-        Just up ->
+      let mbAuthStringToken = case (explicitAuth, uriDerivedAuth) of
+            (Just (Right token), _) -> Just $ Right token
+            (Just (Left (uname, passwd)), _) -> Just $ Left (uname ++ ":" ++ passwd)
+            (_, Just a) -> Just $ Left a
+            (_, _) -> Nothing
```

Comment:
Not happy with catch-alls here, they could (in theory) cause problems in the future. Is there a reason why we should not put `Nothing` there?